spyder/widgets/editor.py

```diff
@@ -274,6 +274,70 @@ def breakpoints_changed(self):
             self.save_breakpoints.emit(self.filename, repr(breakpoints))
 
 
+class TabSwitcherWidget(QListWidget):
+    """Show tabs in mru order and change between them."""
+    def __init__(self, parent, stack_history, tabs):
+        QListWidget.__init__(self, parent)
+        self.setWindowFlags(Qt.SubWindow | Qt.FramelessWindowHint)
+
+        self.editor = parent
+        self.stack_history = stack_history
+        self.tabs = tabs
+
+        self.setSelectionMode(QListWidget.SingleSelection)
+        self.itemActivated.connect(self.item_selected)
+
+        self.id_list = []
+        self.load_data()
+        size = CONF.get('main', 'completion/size')
+        self.resize(*size)
+        self.set_dialog_position()
+        self.setCurrentRow(0)
+
+    def load_data(self):
+        """Fill ListWidget with the tabs texts.
+
+        Add elements in inverse order of stack_history.
+        """
+        # Fill a list to track ids and corresponding tab index
+        self.id_list = [id(self.tabs.widget(i))
+                        for i in range(self.tabs.count())]
+
+        for index in self.stack_history[::-1]:
+            text = self.tabs.tabText(self.id_list.index(index))
+            item = QListWidgetItem(ima.icon('FileIcon'), text)
+            self.addItem(item)

```

Comment:
As discussed @rlaverde lets move all the stack history logic into a separate class (StackHistory)